### PR TITLE
updates for treatment of topography

### PR DESCRIPTION
**Pull request by mt5555**
Several updates related how the NH dycore handles topography, using the Held-Suarez test case

              read topo files for standalone HOMME
              topo smoothing code moved to standalone HOMME
              different pressure gradients (disabled by default)
              different forms fo theta advection (disabled by default)
              tweaks to hypervis ref states for HS with topo
              debugged and turned on heating
              dpnh_dp_i scaling by dt
              make H and NH  DCMIP2016 forcing consistent

### CODE REVIEW

**Comment by oksanaguba** [comment]
What is the difference between hstopo2 and hstopo branches (asking since i was running hstopo before)?  I am still reading diffs but wonder if hs+topo test could be added to the suite?

File: components/homme/src/prim_movie_mod.F90 [hunk 1]
```diff
@@ -210,6 +210,9 @@ subroutine prim_movie_init(elem, par, hvcoord,tl)
     call nf_variable_attributes(ncdf, 'u', 'longitudinal wind component','meters/second')
     call nf_variable_attributes(ncdf, 'v', 'latitudinal wind component','meters/second')
     call nf_variable_attributes(ncdf, 'T', 'Temperature','degrees kelvin')
+#ifdef _PRIM
```
Comment (oksanaguba) [inline components/homme/src/prim_movie_mod.F90]: Below other vars, like hybrid coefficients, are not ifdef-ed with _PRIM.

File: components/homme/src/test_src/held_suarez_mod.F90 [hunk 2]
```diff
@@ -252,9 +272,22 @@ subroutine hs0_init_state(elem, hybrid, hvcoord,nets,nete,Tinit)
...
 #endif
 
-       elem(ie)%state%v(:,:,:,:,n0) =0.0D0
-       elem(ie)%state%v(:,:,:,:,nm1)=elem(ie)%state%v(:,:,:,:,n0)
-       elem(ie)%state%v(:,:,:,:,np1)=elem(ie)%state%v(:,:,:,:,n0)
+
+       ! initialize surface pressure to be consistent with topo
+       ! use 11.3pa per meter 
+       elem(ie)%state%ps_v(:,:,n0) = elem(ie)%state%ps_v(:,:,n0) - &
+            11.3*elem(ie)%state%phis(:,:)/g
```
Comment (oksanaguba) [inline components/homme/src/test_src/held_suarez_mod.F90]: For const T it seems that analytical formula is p(z)=C exp(-g z / R_dry T) for dry air and C obtained from the top pressure...

File: components/homme/src/theta-l/element_state.F90 [hunk 3]
```diff
@@ -69,6 +69,7 @@ module element_state
     real (kind=real_kind) :: FQ(np,np,nlev,qsize_d)                ! tracer forcing
     real (kind=real_kind) :: FM(np,np,3,nlev)                      ! momentum forcing
     real (kind=real_kind) :: FT(np,np,nlev)                        ! temperature forcing
+    real (kind=real_kind) :: FPHI(np,np,nlevp)                     ! PHI (NH) forcing
```
Comment (oksanaguba) [inline components/homme/src/theta-l/element_state.F90]: should this be wrapped with #ifndef CAM?

Comment (oksanaguba) [reply components/homme/src/theta-l/element_state.F90]: minor point: FPHI at the bottom is always 0, so, for alignment/vectorization, would it be better to keep is of size nlev?

File: components/homme/src/theta-l/element_state.F90 [hunk 4]
```diff
@@ -69,6 +69,7 @@ module element_state
...
     real (kind=real_kind) :: FQps(np,np)                   ! forcing of FQ on ps_v
```
Comment (oksanaguba) [inline components/homme/src/theta-l/element_state.F90]: minor point: FPHI at the bottom is always 0, so, for alignment/vectorization, would it be better to keep is of size nlev?

File: components/homme/src/theta-l/prim_advance_mod.F90 [hunk 5]
```diff
@@ -464,6 +509,7 @@ subroutine applyCAMforcing_dynamics(elem,hvcoord,np1,dt,nets,nete)
   do ie=nets,nete
 
      elem(ie)%state%vtheta_dp(:,:,:,np1) = elem(ie)%state%vtheta_dp(:,:,:,np1) + dt*elem(ie)%derived%FT(:,:,:)
+     elem(ie)%state%phinh_i(:,:,1:nlev,np1) = elem(ie)%state%phinh_i(:,:,1:nlev,np1) + dt*elem(ie)%derived%FPHI(:,:,1:nlev)
```
Comment (oksanaguba) [inline components/homme/src/theta-l/prim_advance_mod.F90]: i assume you ran dcmip16 tests with forcings on phi. should i could run them for ftype4?

File: components/homme/src/theta-l/prim_advance_mod.F90 [hunk 6]
```diff
@@ -433,21 +463,36 @@ subroutine convert_thermo_forcing(elem,hvcoord,n0,n0q,dt,nets,nete)
...
+        phi_n1(:,:,k)=phi_n1(:,:,k+1) + Rgas*vthn1(:,:,k)*exner(:,:,k)/pnh(:,:,k)
+     enddo
 
      !finally, compute difference for FT
      ! this method is using new dp, new exner, new-new r*, new t
      elem(ie)%derived%FT(:,:,:) = &
          (vthn1 - elem(ie)%state%vtheta_dp(:,:,:,n0))/dt
 
+     elem(ie)%derived%FPHI(:,:,:) = &
+         (phi_n1 - elem(ie)%state%phinh_i(:,:,:,n0))/dt
```
Comment (oksanaguba) [inline components/homme/src/theta-l/prim_advance_mod.F90]: i am still confused why phi is using forcing and w does not. if one of dyn. eqns, EOS, is used to define forcing on phi, why not use other eqns. for forcing for w?

File: components/homme/src/theta-l/prim_advance_mod.F90 [hunk 7]
```diff
@@ -635,79 +677,71 @@ subroutine advance_hypervis(elem,hvcoord,hybrid,deriv,nt,nets,nete,dt2,eta_ave_w
...
+           !add ref state back
+           elem(ie)%state%vtheta_dp(:,:,k,nt)=elem(ie)%state%vtheta_dp(:,:,k,nt)+&
+                theta_ref(:,:,k,ie)
+           elem(ie)%state%phinh_i(:,:,k,nt)=elem(ie)%state%phinh_i(:,:,k,nt)+&
+                phi_ref(:,:,k,ie)
+           elem(ie)%state%dp3d(:,:,k,nt)=elem(ie)%state%dp3d(:,:,k,nt)+&
+                dp_ref(:,:,k,ie)
+           
+        enddo
+
```
Comment (oksanaguba) [inline components/homme/src/theta-l/prim_advance_mod.F90]: it seems the 2 big blocks above have only formatting diffs.

**Comment by oksanaguba** [comment]
commit for dp3d output was added but since i had an outdated branch at the time, merge commit showed up, too. i can revert 2 commits and redo the dp3d commit if it is better.

**Comment by mt5555** [comment]
merged to next

**Pull request merged**